analyze_test.go

```diff
@@ -2427,3 +2427,88 @@ func TestAnalyze_headers_more(t *testing.T) {
 		})
 	}
 }
+
+//nolint:funlen
+func TestAnalyze_matchFn(t *testing.T) {
+	t.Parallel()
+	// define a test matchFn so that "state" directive has different definition in OSS and matchFn
+	testDirectiveMap := map[string][]uint{
+		"state":    {ngxAnyConf | ngxConfTake2},
+		"test_dir": {ngxAnyConf | ngxConfTake1},
+	}
+	testMatchFn := func(directive string) ([]uint, bool) {
+		masks, matched := testDirectiveMap[directive]
+		return masks, matched
+	}
+
+	testcases := map[string]struct {
+		stmt    *Directive
+		ctx     blockCtx
+		wantErr bool
+	}{
+		"matched only by matchFn": {
+			&Directive{
+				Directive: "test_dir",
+				Args:      []string{"args1"},
+				Line:      5,
+			},
+			blockCtx{"http", "upstream"},
+			false,
+		},
+		"matched by matchFn and default map case1": {
```

Comment:
Which one should provide the mask that matches? Is it the built-in map rather than the matchfn or the other way around? how can the test prove this?

Reply:
case1 is matchFn provides it, default map doesn't and case2 is default map provides, matchFn doesn't. I added some annotations to it. Thanks for pointing it out!